console/src/state/async_ops.rs

```diff
@@ -0,0 +1,321 @@
+    }
+}
+
+impl AsyncOp {
+    pub(crate) fn id(&self) -> u64 {
+        self.id
+    }
+
+    pub(crate) fn parent_id(&self) -> &str {
+        &self.parent_id
+    }
+
+    pub(crate) fn resource_id(&self) -> u64 {
+        self.resource_id
+    }
+
+    pub(crate) fn task_id(&self) -> Option<u64> {
+        self.stats.task_id
+    }
+
+    pub(crate) fn task_id_str(&self) -> &str {
+        &self.stats.task_id_str
+    }
+
+    pub(crate) fn source(&self) -> &str {
+        &self.source
+    }
+
+    pub(crate) fn total(&self, since: SystemTime) -> Duration {
+        self.stats
+            .total
+            .unwrap_or_else(|| since.duration_since(self.stats.created_at).unwrap())
+    }
+
+    pub(crate) fn busy(&self, since: SystemTime) -> Duration {
+        if let (Some(last_poll_started), None) =
+            (self.stats.last_poll_started, self.stats.last_poll_ended)
+        {
+            let current_time_in_poll = since.duration_since(last_poll_started).unwrap();
+            return self.stats.busy + current_time_in_poll;
+        }
+        self.stats.busy
+    }
+
+    pub(crate) fn idle(&self, since: SystemTime) -> Duration {
+        self.stats
+            .idle
+            .unwrap_or_else(|| self.total(since) - self.busy(since))
+    }
+
+    pub(crate) fn total_polls(&self) -> u64 {
+        self.stats.polls
+    }
+
+    pub(crate) fn dropped(&self) -> bool {
+        self.stats.total.is_some()
+    }
+
+    pub(crate) fn formatted_attributes(&self) -> &[Vec<Span<'static>>] {
+        &self.stats.formatted_attributes
+    }
+}
+
+impl AsyncOpStats {
+    fn from_proto(
+        pb: proto::async_ops::Stats,
+        meta: &Metadata,
+        styles: &view::Styles,
+        strings: &mut intern::Strings,
+    ) -> Self {
+        let mut pb = pb;
+
+        let mut attributes = pb
+            .attributes
+            .drain(..)
+            .filter_map(|pb| {
+                let field = pb.field?;
+                let field = Field::from_proto(field, meta, strings)?;
+                Some(Attribute {
+                    field,
+                    unit: pb.unit,
+                })
+            })
+            .collect::<Vec<_>>();
+
+        let created_at = pb
+            .created_at
+            .expect("async op span was never created")
+            .try_into()
+            .unwrap();
+
+        let dropped_at: Option<SystemTime> = pb.dropped_at.map(|v| v.try_into().unwrap());
+        let total = dropped_at.map(|d| d.duration_since(created_at).unwrap());
+
+        let poll_stats = pb.poll_stats.expect("task should have poll stats");
+        let busy = poll_stats.busy_time.map(pb_duration).unwrap_or_default();
+        let idle = total.map(|total| total - busy);
+        let formatted_attributes = Attribute::make_formatted(styles, &mut attributes);
+        let task_id = pb.task_id.map(|id| id.id);
+        let task_id_str = strings.string(
+            task_id
+                .map(|tid| format!("{}", tid))
```

Comment:
nit: could be 
```suggestion
                .map(u64::to_string)
```
which might be very slightly more efficient (but not a big deal)